### ISSUE TITLE
formatter: tailwindcss

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#16542]
https://github.com/tailwindlabs/prettier-plugin-tailwindcss is a AST transform plugin, which does not fit into our pipeline.

This is my proposed plan via Claude:

# Tailwind CSS Integration for oxfmt - Implementation Plan

## Goal
Integrate Tailwind CSS class sorting into oxfmt by detecting class attributes during formatting, collecting them, batch transforming via prettier-plugin-tailwindcss, and then applying the sorted results.

## User Requirements
- ✅ Only sort specific attributes (class, className) - matching prettier-plugin-tailwindcss behavior
- ✅ Modify prettier-plugin-tailwindcss directly in `/Users/boshen/github/prettier-plugin-tailwindcss` and pnpm link for POC
- ✅ Fail silently if sorting fails, keeping original classes
- ✅ Reuse prettier-plugin-tailwindcss's detection logic for auto-enabling

## Architecture Overview

**Two-Pass Design**:
1. **IR Generation Phase**: Detect class attributes, insert placeholders (u32 IDs), collect original strings in thread-safe HashMap
2. **Printing Phase**: Batch-invoke JS callback to sort all classes, replace placeholders with sorted strings

**Why This Works**:
- IR remains immutable (no structural changes after creation)
- Single JS callback per file (minimal boundary crossing overhead)
- Single-threaded design (oxfmt formats one file at a time per thread)
- Fails gracefully (original classes preserved on error)

---

### Step 1: Add IR Placeholder Variant

**File**: `/Users/boshen/oxc/oxc/crates/oxc_formatter/src/formatter/format_element/mod.rs`

**Location**: Lines 55-90 (FormatElement enum)

Add new variant before the closing brace:
```rust
pub enum FormatElement<'a> {
    Space,
    HardSpace,
    Line(LineMode),
    ExpandParent,
    Token { text: &'static str },
    Text { text: &'a str, width: TextWidth },
    LineSuffixBoundary,
    Interned(Interned<'a>),
    BestFitting(BestFittingElement<'a>),
    Tag(Tag),

    /// Placeholder for Tailwind CSS class string to be sorted.
    /// The u32 is an ID that maps to the original class string in TailwindClassMap.
    /// During printing, this will be replaced with the sorted class string.
    TailwindPlaceholder(u32),
}
```

**Location**: Lines 92-107 (Debug impl)

Add new match arm before closing brace:
```rust
impl std::fmt::Debug for FormatElement<'_> {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            FormatElement::Space | FormatElement::HardSpace => fmt.write_str("Space"),
            FormatElement::Line(mode) => fmt.debug_tuple("Line").field(mode).finish(),
            FormatElement::ExpandParent => fmt.write_str("ExpandParent"),
            FormatElement::Token { text } => fmt.debug_tuple("Token").field(text).finish(),
            FormatElement::Text { text, .. } => fmt.debug_tuple("Text").field(text).finish(),
            FormatElement::LineSuffixBoundary => fmt.write_str("LineSuffixBoundary"),
            FormatElement::BestFitting(best_fitting) => {
                fmt.debug_tuple("BestFitting").field(&best_fitting).finish()
            }
            FormatElement::Interned(interned) => fmt.debug_list().entries(&**interned).finish(),
            FormatElement::Tag(tag) => fmt.debug_tuple("Tag").field(tag).finish(),
            FormatElement::TailwindPlaceholder(id) => {
                fmt.debug_tuple("TailwindPlaceholder").field(id).finish()
            }
        }
    }
}
```

### Step 2: Create Class Storage (Single-Threaded)

**File**: `/Users/boshen/oxc/oxc/crates/oxc_formatter/src/formatter/context.rs`

**Location**: After imports, before FormatContext struct

Add imports:
```rust
use std::collections::HashMap;
```

Add new struct:
```rust
/// Storage for Tailwind CSS classes awaiting sorting.
/// Single-threaded - no need for Arc/Mutex since formatting happens on one thread per file.
#[derive(Clone, Default)]
pub struct TailwindClassMap {
    map: HashMap<String, u32>,
    next_id: u32,
}

impl TailwindClassMap {
    pub fn new() -> Self {
        Self::default()
    }

    /// Insert class string, return unique ID (deduplicates automatically)
    pub fn insert(&mut self, class_str: String) -> u32 {
        if let Some(&existing_id) = self.map.get(&class_str) {
            return existing_id;
        }

        let new_id = self.next_id;
        self.next_id += 1;
        self.map.insert(class_str, new_id);
        new_id
    }

    /// Drain all classes for sorting (returns Vec<(id, string)>)
    pub fn drain(&mut self) -> Vec<(u32, String)> {
        let drained: Vec<_> = self.map.drain().map(|(s, id)| (id, s)).collect();
        self.next_id = 0;
        drained
    }

    pub fn is_empty(&self) -> bool {
        self.map.is_empty()
    }
}
```

**Location**: FormatContext struct definition

Add field to `FormatContext`:
```rust
pub struct FormatContext<'ast> {
    options: FormatOptions,
    source_text: SourceText<'ast>,
    source_type: SourceType,
    comments: Comments<'ast>,
    cached_elements: FxHashMap<Span, FormatElement<'ast>>,
    allocator: &'ast Allocator,
    embedded_formatter: Option<EmbeddedFormatter>,

    /// Storage for Tailwind CSS class strings to be sorted
    pub(crate) tailwind_class_map: Option<TailwindClassMap>,
}
```

**Location**: FormatContext::new() method

Update constructor signature and initialization:
```rust
pub fn new(
    source_text: &'ast str,
    source_type: SourceType,
    comments: &'ast [Comment],
    allocator: &'ast Allocator,
    options: FormatOptions,
    embedded_formatter: Option<EmbeddedFormatter>,
    tailwind_class_map: Option<TailwindClassMap>,  // NEW parameter
) -> Self {
    let source_text = SourceText::new(source_text);
    Self {
        options,
        source_text,
        source_type,
        comments: Comments::new(source_text, comments),
        allocator,
        cached_elements: FxHashMap::default(),
        embedded_formatter,
        tailwind_class_map,  // NEW field
    }
}
```

**Location**: FormatContext::dummy() method

Update dummy constructor:
```rust
pub(crate) fn dummy(allocator: &'ast Allocator) -> Self {
    Self {
        options: FormatOptions::default(),
        source_text: SourceText::new(""),
        source_type: SourceType::default(),
        comments: Comments::new(SourceText::new(""), &[]),
        allocator,
        cached_elements: FxHashMap::default(),
        embedded_formatter: None,
        tailwind_class_map: None,  // NEW field
    }
}
```

### Step 3: Detection Logic for JSX Attributes

**File**: `/Users/boshen/oxc/oxc/crates/oxc_formatter/src/write/jsx/mod.rs`

**Location**: JSXAttribute FormatWrite implementation (around lines 320-327)

Find the current implementation:
```rust
impl<'a> FormatWrite<'a> for AstNode<'a, JSXAttribute<'a>> {
    fn write(&self, f: &mut Formatter<'_, 'a>) {
        write!(f, self.name());
        if let Some(value) = &self.value() {
            write!(f, ["=", value]);
        }
    }
}
```

Replace with:
```rust
impl<'a> FormatWrite<'a> for AstNode<'a, JSXAttribute<'a>> {
    fn write(&self, f: &mut Formatter<'_, 'a>) {
        write!(f, self.name());

        if let Some(value) = &self.value() {
            // Check if this is a class/className attribute with static string value
            let is_class_attr = matches!(
                self.name(),
                JSXAttributeName::Identifier(id)
                if id.name == "class" || id.name == "className"
            );

            if is_class_attr {
                // Only handle static string literals
                if let JSXAttributeValue::StringLiteral(string_lit) = value {
                    // Check if Tailwind sorting is enabled
                    if let Some(map) = f.context_mut().tailwind_class_map.as_mut() {
                        let class_string = string_lit.value.to_string();
                        let placeholder_id = map.insert(class_string);

                        // Get the quote character from source
                        let source = string_lit.span.source_text(f.context().source_text());
                        let quote = source.chars().next().unwrap_or('"');

                        // Write: = + quote + placeholder + quote
                        write!(f, [
                            token("="),
                            token(if quote == '\'' { "'" } else { "\"" }),
                        ]);
                        f.write_element(FormatElement::TailwindPlaceholder(placeholder_id));
                        write!(f, [
                            token(if quote == '\'' { "'" } else { "\"" }),
                        ]);
                        return;
                    }
                }
            }

            // Default path: write value normally
            write!(f, ["=", value]);
        }
    }
}
```

**Key Points**:
- Only processes `class` or `className` attributes
- Only handles static `StringLiteral` values (not expressions or template literals)
- Preserves quote style (single vs double quotes)
- Falls back to normal formatting if Tailwind map is not present
- Requires mutable access to context via `context_mut()`

### Step 4: Add NAPI Callback Infrastructure

**File**: `/Users/boshen/oxc/oxc/apps/oxfmt/src/core/external_formatter.rs`

Add new types:
```rust
pub type JsSortTailwindClassesCb = ThreadsafeFunction<
    FnArgs<String>,      // JSON input
    Promise<String>,     // JSON output
    FnArgs<String>,
    Status,
    false,
>;

type TailwindSorterCallback = Arc<
    dyn Fn(Vec<(u32, String)>) -> Result<HashMap<u32, String>, String>
    + Send + Sync
>;
```

Update `ExternalFormatter` struct to include `sort_tailwind_classes: Option<TailwindSorterCallback>`.

Add wrapper function `wrap_sort_tailwind()` for JSON marshaling.

### Step 5: Update NAPI Entry Point

**File**: `/Users/boshen/oxc/oxc/apps/oxfmt/src/main_napi.rs`

Add optional parameter `sort_tailwind_cb: Option<JsSortTailwindClassesCb>` to `format()` function.

Pass callback to `ExternalFormatter::new()`.

### Step 6: Modify Printer to Invoke Callback

**File**: `/Users/boshen/oxc/oxc/crates/oxc_formatter/src/formatter/printer/mod.rs`

Update `Printer::print()`:
- Accept `tailwind_class_map: Option<&TailwindClassMap>` parameter
- Before printing, drain class map and invoke callback if present
- Store sorted results in `PrinterState`

Add to `PrinterState`:
```rust
sorted_tailwind_classes: Option<HashMap<u32, String>>,
```

Handle `TailwindPlaceholder` in `print_element()`:
- Look up sorted string by ID
- Print the sorted string
- Gracefully handle missing IDs (print empty, log warning)

#### 7a. Create Batch Sorting Function

**File**: `/Users/boshen/github/prettier-plugin-tailwindcss/src/batch-sort.ts` (NEW FILE)

Create new file with batch sorting logic:
```typescript
import { getTailwindConfig } from './config.js'
import { sortClasses } from './sorting.js'
import type { TransformerEnv } from './types'

/**
 * Batch sort Tailwind CSS classes from oxfmt NAPI binding.
 *
 * @param classes - Array of [id, classString] tuples (JSON parsed)
 * @param options - Prettier-compatible options (must include filepath, plugins, etc.)
 * @returns Object mapping id -> sortedClassString
 */
export async function batchSortTailwindClasses(
  classes: Array<[number, string]>,
  options: any
): Promise<Record<number, string>> {
  // Initialize Tailwind context once per batch
  const context = await getTailwindConfig(options)

  const env: TransformerEnv = {
    context,
    matcher: null as any, // Not used for pure sorting
    parsers: {},
    options,
  }

  const result: Record<number, string> = {}

  for (const [id, classString] of classes) {
    try {
      const sorted = sortClasses(classString, {
        env,
        removeDuplicates: !options.tailwindPreserveDuplicates,
        collapseWhitespace: !options.tailwindPreserveWhitespace ? { start: true, end: true } : false,
      })
      result[id] = sorted
    } catch (err) {
      console.warn(`[prettier-plugin-tailwindcss] Failed to sort classes for ID ${id}:`, err)
      result[id] = classString // Fallback to original
    }
  }

  return result
}
```

#### 7b. Create Detection Function

**File**: `/Users/boshen/github/prettier-plugin-tailwindcss/src/detect.ts` (NEW FILE)

Create detection function that checks if Tailwind CSS is available:
```typescript
import { existsSync } from 'fs'
import { join, dirname } from 'path'
import { getTailwindConfig } from './config.js'

/**
 * Detect if Tailwind CSS is configured for the current project.
 *
 * This function checks:
 * 1. If tailwind.config.js/ts/cjs/mjs exists
 * 2. If Tailwind CSS package is installed
 * 3. If getTailwindConfig can successfully initialize
 *
 * @param options - Prettier options (must include filepath)
 * @returns true if Tailwind CSS is available and configured
 */
export async function shouldEnableTailwindSort(options: any): Promise<boolean> {
  try {
    // Must have a filepath to detect project root
    if (!options.filepath) {
      return false
    }

    const projectRoot = findProjectRoot(options.filepath)

    // Check for Tailwind config file
    const configFiles = [
      'tailwind.config.js',
      'tailwind.config.ts',
      'tailwind.config.cjs',
      'tailwind.config.mjs',
    ]

    const hasConfigFile = configFiles.some(file =>
      existsSync(join(projectRoot, file))
    )

    if (!hasConfigFile) {
      return false
    }

    // Try to initialize Tailwind context
    const context = await getTailwindConfig(options)

    // Check if context has getClassOrder function (core requirement)
    return typeof context?.getClassOrder === 'function'
  } catch (err) {
    // Silent fail - just means Tailwind isn't available
    return false
  }
}

/**
 * Find project root by looking for package.json
 */
function findProjectRoot(filepath: string): string {
  let dir = dirname(filepath)

  while (dir !== dirname(dir)) { // Stop at filesystem root
    if (existsSync(join(dir, 'package.json'))) {
      return dir
    }
    dir = dirname(dir)
  }

  return dirname(filepath) // Fallback to file's directory
}
```

#### 7c. Export Functions

**File**: `/Users/boshen/github/prettier-plugin-tailwindcss/src/index.ts`

Add exports at the end of the file:
```typescript
// ... existing exports ...

// Export functions for oxfmt integration
export { batchSortTailwindClasses } from './batch-sort.js'
export { shouldEnableTailwindSort } from './detect.js'
```

**Important Notes**:
- Detection function is async because it needs to load Tailwind config
- Detection should be called once per oxfmt session, not per file
- Batch sorting is called once per file after IR generation
- Both functions handle errors gracefully by returning false/empty results

### Step 8: Add Configuration Option

**File**: `/Users/boshen/oxc/oxc/crates/oxc_formatter/src/options.rs`

**Location**: FormatOptions struct definition

Add field to FormatOptions:
```rust
#[derive(Debug, Default, Clone)]
pub struct FormatOptions {
    // ... existing fields like indent_style, indent_width, line_width, quotes, etc. ...

    /// Enable Tailwind CSS class sorting (experimental).
    /// When enabled, class and className attributes will be sorted using prettier-plugin-tailwindcss.
    pub tailwindcss: bool,
}
```

**Location**: Display implementation

Update the Display impl to show the new option:
```rust
impl std::fmt::Display for FormatOptions {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // ... existing fields ...
        writeln!(f, "Tailwind CSS sorting: {}", self.tailwindcss)?;
        Ok(())
    }
}
```

**Initialization Flow**:

1. **JS Entry Point** (`apps/oxfmt/index.js` or similar):
   ```javascript
   import { shouldEnableTailwindSort, batchSortTailwindClasses } from 'prettier-plugin-tailwindcss'
   import { format as nativeFormat } from './binding' // NAPI

   export async function format(args, options = {}) {
     // Detect if Tailwind CSS should be enabled
     const enableTailwind = options.filepath
       ? await shouldEnableTailwindSort({ filepath: options.filepath })
       : false

     const formatOptions = {
       ...options,
       tailwindcss: enableTailwind, // Pass to Rust via NAPI
     }

     // ... rest of setup ...
   }
   ```

2. **Rust Side** (`crates/oxc_formatter/src/lib.rs`):
   ```rust
   pub fn format_with_options(&self, program: &Program, options: FormatOptions) -> Formatted {
       let tailwind_class_map = if options.tailwindcss {
           Some(TailwindClassMap::new())
       } else {
           None
       };

       let context = FormatContext::new(
           program.source_text,
           program.source_type,
           &program.comments,
           self.allocator,
           options,
           embedded_formatter,
           tailwind_class_map,
       );

       // ... continue formatting ...
   }
   ```

3. **During Formatting**: If `tailwind_class_map` exists in context, JSX attribute detection logic inserts placeholders

4. **Before Printing**: Drain the map and invoke JS callback to sort classes

This allows automatic detection while still giving users manual control via explicit `tailwindcss: true/false` option.

---

### Rust Files (oxfmt)

1. **`/Users/boshen/oxc/oxc/crates/oxc_formatter/src/formatter/format_element/mod.rs`**
   - Add `TailwindPlaceholder(u32)` to FormatElement enum
   - Update Debug impl
   - ~10 lines added

2. **`/Users/boshen/oxc/oxc/crates/oxc_formatter/src/formatter/context.rs`**
   - Add TailwindClassMap struct (~40 lines)
   - Add field to FormatContext
   - Update constructor and dummy method

3. **`/Users/boshen/oxc/oxc/crates/oxc_formatter/src/write/jsx/mod.rs`**
   - Modify JSXAttribute::write() implementation
   - Add class/className detection logic
   - Insert placeholders for Tailwind classes
   - ~40 lines modified/added

4. **`/Users/boshen/oxc/oxc/crates/oxc_formatter/src/options.rs`**
   - Add `tailwindcss: bool` field to FormatOptions
   - Update Display impl
   - ~3 lines added

5. **`/Users/boshen/oxc/oxc/crates/oxc_formatter/src/formatter/printer/mod.rs`**
   - Modify Printer::print() to invoke callback
   - Add sorted_tailwind_classes field to PrinterState
   - Handle TailwindPlaceholder in print_element()
   - ~50 lines added/modified

6. **`/Users/boshen/oxc/oxc/apps/oxfmt/src/core/external_formatter.rs`**
   - Add JsSortTailwindClassesCb type
   - Add TailwindSorterCallback type
   - Update ExternalFormatter struct
   - Add wrap_sort_tailwind() function
   - ~60 lines added

7. **`/Users/boshen/oxc/oxc/apps/oxfmt/src/main_napi.rs`**
   - Add sort_tailwind_cb parameter to format() function
   - Pass callback to ExternalFormatter::new()
   - ~5 lines modified

8. **`/Users/boshen/oxc/oxc/crates/oxc_formatter/src/lib.rs`**
   - Update format() to conditionally create TailwindClassMap
   - Pass map to FormatContext::new()
   - ~10 lines modified

### JavaScript Files (prettier-plugin-tailwindcss)

9. **`/Users/boshen/github/prettier-plugin-tailwindcss/src/batch-sort.ts`** (NEW)
   - Create batchSortTailwindClasses() function
   - ~40 lines

10. **`/Users/boshen/github/prettier-plugin-tailwindcss/src/detect.ts`** (NEW)
    - Create shouldEnableTailwindSort() function
    - Create findProjectRoot() helper
    - ~60 lines

11. **`/Users/boshen/github/prettier-plugin-tailwindcss/src/index.ts`**
    - Export batch-sort and detect functions
    - ~2 lines added

12. **`/Users/boshen/oxc/oxc/apps/oxfmt/index.js`** (or JS wrapper)
    - Call shouldEnableTailwindSort() during initialization
    - Implement JS callback that calls batchSortTailwindClasses()
    - Pass callback to NAPI format() function
    - ~40 lines added/modified

**Total**: ~12 files, ~360 lines of code

---

### MVP Scope (What We Handle)
- ✅ Static string literals in JSX: `<div className="z-10 p-4 m-2 bg-red-500" />`
- ✅ HTML class attributes: `<div class="flex items-center" />`

### Out of Scope (Future)
- ❌ Template literals with expressions: `className={\`foo ${bar} baz\`}`
- ❌ Conditional classes: `className={cond ? "foo" : "bar"}`
- ❌ Spread attributes: `<div {...props} className="foo" />`
- ❌ Framework-specific directives: `:class`, `[ngClass]`, etc.

---

## Error Handling Strategy

1. **JSON marshaling fails**: Log error, return empty map (all placeholders print empty)
2. **JS callback fails**: Catch, log warning, use original strings
3. **Missing ID during printing**: Log warning, print empty string
4. **Tailwind config not found**: prettier-plugin-tailwindcss handles detection, fails gracefully
5. **No crash guarantee**: All error paths must preserve formatting operation

---

### Unit Tests
- `TailwindClassMap` insert/drain operations
- JSON serialization roundtrip
- IR generation with placeholders
- Printer placeholder handling

### Integration Tests
- Format fixture files with class attributes
- Verify sorted output matches prettier-plugin-tailwindcss
- Test error cases (invalid config, missing dependencies)
- Verify existing tests still pass

### NAPI Tests
- JavaScript callback invocation
- JSON payload validation
- Async error handling

---

## Success Criteria

✅ Sorts static class/className strings in JSX
✅ Integrates with prettier-plugin-tailwindcss via NAPI
✅ Fails gracefully (no crashes)
✅ Preserves all existing formatting
✅ Thread-safe for parallel formatting
✅ <5% performance overhead when disabled

**Comment by leaysgur** [comment]
Supported by v0.22 ✨ 
https://github.com/oxc-project/oxc/releases/tag/apps_v1.37.0

**Comment by stunaz** [comment]
I am in a monorepo, it doesn't work, it removes all classes  and leave empty string

**Comment by Dunqing** [comment]
> I am in a monorepo, it doesn't work, it removes all classes and leave empty string

Sorry for causing this trouble. Is this a duplicate of https://github.com/oxc-project/oxc/issues/17725?

We've identified this significant bug and have already fixed it, but it has not been released yet. It will be released next Monday.

**Comment by bitttttten** [comment]
This is great, and thanks so much for adding it! We installed it on our monorepo and our formatting went from 2s to 9s, and also it requires node_modules to be installed. One thing we liked with oxfmt if that we could run something like `pnpx oxfmt@0.24.0 --check .` and therefore this was real fast on our CI at catching formatting issues earlier, since it did not rely on installing node modules.

Two questions:

- Is there a plan to not use the prettier plugin to organise sorting class names? And just to do it without any deps?
- Are we doing something wrong that it falls back to running "prettier-plugin-tailwindcss"? i.e. we are hitting this code branch `prettierTailwindcss = await import("prettier-plugin-tailwindcss");` [from this PR](https://github.com/oxc-project/oxc/pull/16990/changes#diff-8352cb227d4cd546557c88c45cc52deff344c641046e73520707473ec7d51375R136).

Thanks again for the work here and still sub 10s is great for our monoreop, we were just surprised to notice it was 5 times slower (even tho again, 10s is not slow!).

🙇

**Comment by Stybo** [comment]
Is there a linter for this using Oxlint?

**Comment by camc314** [comment]
> Is there a linter for this using Oxlint?

@Stybo yes

please look at https://github.com/camc314/oxfmt-oxlint-tailwindcss

**Comment by Stybo** [comment]
@camc314 Sorry found this first: https://github.com/sergioazoc/oxlint-tailwindcss

**Comment by camc314** [comment]
@Stybo good glad you found something that works - either case works, the repo I posted also demo's how to use any JS plugin with oxlint - thanks!

**Comment by Stybo** [comment]
@camc314 True! it just has some more edge cases and config options so i prefer that, but nice plugin for people that don't care those things!

**Comment by Stybo** [comment]
@camc314 Using your plugin now because the other one has class order conflicts with the formatter of oxfmt.

**Comment by camc314** [comment]
> @camc314 Using your plugin now because the other one has class order conflicts with the formatter of oxfmt.

just disable the lint rule that is doing class sorting then - should fix it.

**Comment by Stybo** [comment]
> > [@camc314](https://github.com/camc314) Using your plugin now because the other one has class order conflicts with the formatter of oxfmt.
> 
> just disable the lint rule that is doing class sorting then - should fix it.

Then it wont error when it is not sorted, defeats a bit of the purpose. I made an issue, hopefully it will be fixed

**Comment by camc314** [comment]
> > > [@camc314](https://github.com/camc314) Using your plugin now because the other one has class order conflicts with the formatter of oxfmt.
> > 
> > 
> > just disable the lint rule that is doing class sorting then - should fix it.
> 
> Then it wont error when it is not sorted, defeats a bit of the purpose. I made an issue, hopefully it will be fixed

but oxfmt should error as there will be a diff before/after formatting
